keanu-project/src/main/java/io/improbable/keanu/distributions/tensors/continuous/TensorSmoothUniform.java

```diff
@@ -0,0 +1,119 @@
+package io.improbable.keanu.distributions.tensors.continuous;
+
+import io.improbable.keanu.vertices.dbltensor.DoubleTensor;
+import io.improbable.keanu.vertices.dbltensor.KeanuRandom;
+
+public class TensorSmoothUniform {
+
+    private TensorSmoothUniform() {
+    }
+
+    public static DoubleTensor sample(int[] shape, DoubleTensor xMin, DoubleTensor xMax, double edgeSharpness, KeanuRandom random) {
+
+        DoubleTensor r1 = random.nextDouble(shape);
+        DoubleTensor r2 = random.nextDouble(shape);
+
+        final DoubleTensor bodyWidth = xMax.minus(xMin);
+        final DoubleTensor shoulderWidth = bodyWidth.times(edgeSharpness);
+        final DoubleTensor rScaled = r1.times(bodyWidth.plus(shoulderWidth)).plusInPlace(xMin.minus(shoulderWidth.div(2)));
+        final DoubleTensor bodyHeight = bodyHeight(shoulderWidth, bodyWidth);
+
+        final DoubleTensor firstConditional = rScaled.getGreaterThanOrEqualToMask(xMin);
+        firstConditional.timesInPlace(rScaled.getLessThanOrEqualToMask(xMax));
+        final DoubleTensor inverseFirstConditional = DoubleTensor.ones(firstConditional.getShape()).minusInPlace(firstConditional);
+
+        final DoubleTensor secondConditional = rScaled.getLessThanMask(xMin);
+        DoubleTensor spillOnToShoulder = xMin.minus(rScaled);
+        DoubleTensor shoulderX = shoulderWidth.minus(spillOnToShoulder);
+        DoubleTensor shoulderDensity = shoulder(shoulderWidth, bodyWidth, shoulderX);
+        DoubleTensor acceptProbability = shoulderDensity.div(bodyHeight);
+
+        final DoubleTensor secondConditionalNestedTrue = secondConditional.times(r2.getLessThanOrEqualToMask(acceptProbability));
+        final DoubleTensor secondConditionalNestedFalse = secondConditional.times(r2.getGreaterThanMask(acceptProbability));
+        final DoubleTensor secondConditionalNestedFalseResult = xMin.minus(shoulderWidth).plus(spillOnToShoulder);
+
+        final DoubleTensor secondConditionalFalse = rScaled.getGreaterThanOrEqualToMask(xMin);
+        spillOnToShoulder = rScaled.minus(xMax);
+        shoulderX = shoulderWidth.minus(spillOnToShoulder);
+        shoulderDensity = shoulder(shoulderWidth, bodyWidth, shoulderX);
+        acceptProbability = shoulderDensity.divInPlace(bodyHeight);
+
+        final DoubleTensor secondConditionalFalseNestedTrue = secondConditionalFalse.times(r2.getLessThanOrEqualToMask(acceptProbability));
+        final DoubleTensor secondConditionalFalseNestedFalse = secondConditionalFalse.times(r2.getGreaterThanMask(acceptProbability));
```

Comment:
Since it's the last time shoulder is used and the last time secondConditionalFalse is used.
```
        final DoubleTensor secondConditionalFalseNestedFalse = secondConditionalFalse.timesInPlace(r2.getGreaterThanMask(acceptProbability));
        final DoubleTensor secondConditionalFalseNestedFalseResult = shoulderWidth.plusInPlace(xMax).minusInPlace(spillOnToShoulder);
```

Reply:
nice, changed